### PR TITLE
OpenPCDet LiDar Object Detection Node

### PR DESCRIPTION
**Pull request by danielrhuynh**
## Overview
Added a ros node for LiDar object detection node trained on nuscenes point clouds.
source repo: https://github.com/open-mmlab/OpenPCDet

**demos:**

https://github.com/WATonomous/wato_monorepo/assets/89366190/542a706f-4143-49e6-ba1d-355c5c492a6b

https://github.com/WATonomous/wato_monorepo/assets/89366190/abb7dc29-588b-4275-a260-8098ffdaac4a

### CODE REVIEW

File: docker/perception/lidar_object_detection/lidar_object_detection_openpcdet.Dockerfile
Comment (Edwardius) [inline docker/perception/lidar_object_detection/lidar_object_detection_openpcdet.Dockerfile:1]: plz use given dockerfile template

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile [hunk 1]
```diff
@@ -32,20 +82,20 @@ WORKDIR /
 RUN apt-get -qq autoremove -y && apt-get -qq autoclean && apt-get -qq clean && \
     rm -rf /root/* /root/.ros /tmp/* /var/lib/apt/lists/* /usr/share/doc/*
 
-################################ Build ################################
+################################ Source #######################################
```
Comment (Edwardius) [inline docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:85]: accident?

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile [hunk 2]
```diff
@@ -7,20 +7,70 @@ WORKDIR ${AMENT_WS}/src
...
+    git zip unzip libssl-dev libcairo2-dev lsb-release libgoogle-glog-dev libgflags-dev libatlas-base-dev libeigen3-dev software-properties-common \
+    build-essential cmake pkg-config libapr1-dev autoconf automake libtool curl libc6 libboost-all-dev debconf libomp5 libstdc++6 \
+    libqt5core5a libqt5xml5 libqt5gui5 libqt5widgets5 libqt5concurrent5 libqt5opengl5 libcap2 libusb-1.0-0 libatk-adaptor neovim \
+    python3-pip python3-setuptools \
+    && apt-get -qq autoremove -y && apt-get -qq autoclean && apt-get -qq clean && \
+    rm -rf /var/lib/apt/lists/* /root/* /root/.ros /tmp/* /usr/share/doc
+
+################################ INSTALL OpenCV with CUDA Support ##############
+WORKDIR /opt
+RUN git clone https://github.com/opencv/opencv.git && \
```
Comment (Edwardius) [inline docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:32]: directly clone from branch using `git clone -b`

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile [hunk 3]
```diff
@@ -7,20 +7,70 @@ WORKDIR ${AMENT_WS}/src
...
+    cd opencv && git checkout 4.5.5
+RUN git clone https://github.com/opencv/opencv_contrib.git && \
```
Comment (Edwardius) [inline docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:34]: directly clone from branch using `git clone -b`

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile (hunk 2 above)
Comment (Edwardius) [inline docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:32]: put clones in the source section (Line 3)

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile (hunk 3 above)
Comment (Edwardius) [reply docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:34]: also put clones in the source section

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile (hunk 2 above)
Comment (Edwardius) [reply docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:32]: source means source code

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile [hunk 4]
```diff
@@ -7,20 +7,70 @@ WORKDIR ${AMENT_WS}/src
...
+ENV PATH /usr/local/cuda/bin:${PATH}
+ENV OpenCV_DIR=/usr/share/OpenCV
+# Install Python dependencies
+RUN pip3 install torch==1.13.1+cu116 torchvision==0.14.1+cu116 torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cu116
+RUN pip3 install torch-scatter -f https://data.pyg.org/whl/torch-1.13.1+cu116.html
+ENV TORCH_CUDA_ARCH_LIST="3.5;5.0;6.0;6.1;7.0;7.5;8.0;8.6+PTX"
+RUN pip3 install spconv-cu116 pyquaternion numpy==1.23 pillow==8.4 mayavi open3d av2
+WORKDIR /home/bolty
+RUN git clone https://github.com/WATonomous/OpenPCDet.git
+RUN cd OpenPCDet && pip3 install -r requirements.txt
```
Comment (Edwardius) [inline docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:65]: reference a specific commit hash, and put in source section

File: modules/docker-compose.perception.yaml [hunk 5]
```diff
@@ -44,8 +44,23 @@ services:
...
+      # New Models to Try
+      - /mnt/wato-drive/perception/bevfusion_pretrain/bevfusion-det.pth:/home/bolty/OpenPCDet/models/bevfusion-det.pth
+      - /mnt/wato-drive/perception/bevfusion_pretrain/lidar-only-det.pth:/home/bolty/OpenPCDet/models/lidar-only-det.pth
+      - /mnt/wato-drive/perception/bevfusion_pretrain/lidar-only-seg.pth:/home/bolty/OpenPCDet/models/lidar-only-seg.pth
+      - /mnt/wato-drive/perception/bevfusion_pretrain/lidar-only-seg.pth:/home/bolty/OpenPCDet/models/swint-nuimages-pretrained.pth
+      - /mnt/wato-drive/perception/pcds/pvrcnn_ros_ckpt.pth:/home/bolty/OpenPCDet/models/pvrcnn_ros_ckpt.pth
+      - /mnt/wato-drive/perception/pcds/pvrcnn_base_ckpt.pth:/home/bolty/OpenPCDet/models/pvrcnn_base_ckpt.pth
+      # Data sets
+      - /mnt/wato-drive2/nuscenes-1.0-mini:/home/bolty/OpenPCDet/data/nuscenes/v1.0-mini
+      - /mnt/wato-drive2/wato-datasets/nuscenes/full-dataset:/home/bolty/OpenPCDet/data/nuscenes/v1.0-trainval
```
Comment (Edwardius) [inline modules/docker-compose.perception.yaml:62]: do we need to mount all of these for running in production? if not, add them to the dev_override

File: modules/docker-compose.perception.yaml [hunk 6]
```diff
@@ -88,4 +103,4 @@ services:
         - "${PERCEPTION_DEPTH_ESTIMATION_IMAGE}:build_main"
       target: deploy
     image: "${PERCEPTION_DEPTH_ESTIMATION_IMAGE}:${TAG}"
-    command: /bin/bash -c "ros2 launch depth_estimation eve.launch.py"
+    command: /bin/bash -c "ros2 launch depth_estimation eve.launch.py"
```
Comment (Edwardius) [inline modules/docker-compose.perception.yaml:106]: wrong package

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+from pcdet.datasets import DatasetTemplate
+from pcdet.models import build_network, load_data_to_gpu
+from pcdet.utils import common_utils
+import torch
+import numpy as np
+from rclpy.node import Node
+import rclpy
+import argparse
+import sys
+sys.path.append("/home/bolty/OpenPCDet")
```
Comment (Edwardius) [inline src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:15]: no absolute paths

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+
+class LidarObjectDetection(Node):
+    def __init__(self):
+        super().__init__('lidar_object_detection')
+        self.declare_parameter(
+            "model_path", "/home/bolty/OpenPCDet/models/transfusion_trained_model.pth")
```
Comment (Edwardius) [inline src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:22]: no absolute paths, best to not have a default because it becomes confusing to know whats setting the parameter

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        self.declare_parameter(
+            "model_config_path", "/home/bolty/OpenPCDet/tools/cfgs/nuscenes_models/transfusion_lidar.yaml")
+        self.declare_parameter("lidar_topic", "/velodyne_points")
+        self.model_path = self.get_parameter("model_path").value
+        self.model_config_path = self.get_parameter("model_config_path").value
+        self.lidar_data = self.get_parameter("lidar_topic").value
```
Comment (Edwardius) [inline src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:28]: best to not have a default because it becomes confusing to know whats setting the parameter

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+        args, cfg = self.parse_config()
+        self.logger = common_utils.create_logger()
+
+        self.demo_dataset = LidarDataset(
+            dataset_cfg=cfg.DATA_CONFIG,
+            class_names=cfg.CLASS_NAMES,
+            training=False,
+            logger=self.logger,
+        )
```
Comment (Edwardius) [inline src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:45]: whats this for?

Comment (Edwardius) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:45]: oh, I would call this a dataloader not a dataset

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+                detection.bbox.size.x = float(box[3])
+                detection.bbox.size.y = float(box[4])
+                detection.bbox.size.z = float(box[5])
+                detected_object = ObjectHypothesisWithPose()
+                detected_object.hypothesis.class_id = str(pred_dicts[0]["pred_labels"][idx])
+                detected_object.hypothesis.score = float(pred_dicts[0]["pred_scores"][idx])
+                detection.results.append(detected_object)
+                detections.detections.append(detection)
+
+        self.viz_publisher.publish(marker_array)
```
Comment (Edwardius) [inline src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:126]: this should be optional inside the config. Better yet, make foxglove directly accept Detection3D

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        )
+        parser.add_argument(
+            "--ckpt", type=str, default=self.model_path, help="specify the pretrained model"
+        )
+
+        args, _ = parser.parse_known_args()
+        cfg_from_yaml_file(args.cfg_file, cfg)
+        return args, cfg
+
+    def create_cloud_xyz32(self, header, points):
```
Comment (Edwardius) [inline src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:145]: is this used?

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+                detection = Detection3D()
+                detection.header = pointcloud_msg.header
+                detection.bbox.center.position.x = float(box[0])
+                detection.bbox.center.position.y = float(box[1])
+                detection.bbox.center.position.z = float(box[2])
+                detection.bbox.size.x = float(box[3])
+                detection.bbox.size.y = float(box[4])
+                detection.bbox.size.z = float(box[5])
+                detected_object = ObjectHypothesisWithPose()
+                detected_object.hypothesis.class_id = str(pred_dicts[0]["pred_labels"][idx])
```
Comment (Edwardius) [inline src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:121]: these labels should be pulled out of the model and stored centrally @danielrhuynh in some sort of label server that publishes `/vision_msgs/VisionInfo.msg`

**Review by Edwardius — CHANGES_REQUESTED** [review]
Awesomeeeee work! Just some minor comments here and there. How big is the Lidar Image?

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile (hunk 1 above)
Comment (danielrhuynh) [reply docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:85]: Yup lol

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile (hunk 2 above)
Comment (danielrhuynh) [reply docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:32]: word, done!

File: docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile (hunk 3 above)
Comment (danielrhuynh) [reply docker/perception/lidar_object_detection/lidar_object_detection.Dockerfile:34]: bet

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 12 above)
Comment (danielrhuynh) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:145]: remnants lol good catch

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 13 above)
Comment (danielrhuynh) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:121]: Not quite sure what this means lol

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 11 above)
Comment (danielrhuynh) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:126]: the marker arrays?

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 13 above)
Comment (danielrhuynh) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:119]: The pred labels?

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 11 above)
Comment (danielrhuynh) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:124]: Think I got it

**Comment by danielrhuynh** [comment]
> Awesomeeeee work! Just some minor comments here and there. How big is the Lidar Image?

Thanks! I'm working on those comments, the image is 900MB

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 11 above)
Comment (Edwardius) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py]: yeah like optionally pulished

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 13 above)
Comment (Edwardius) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:121]: class labels yeah, otherwise we don't know what the class_id actually refers too downstream

Comment (Edwardius) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:121]: eg. how do we know `1` refers to a `car`?

**Comment by Edwardius** [comment]
> > Awesomeeeee work! Just some minor comments here and there. How big is the Lidar Image?
> 
> Thanks! Comments are for the most part addressed, the image is 900MB

oh! We shouldnt be skipping it's build then

**Comment by danielrhuynh** [comment]
> > > Awesomeeeee work! Just some minor comments here and there. How big is the Lidar Image?
> > 
> > 
> > Thanks! Comments are for the most part addressed, the image is 900MB
> 
> oh! We shouldnt be skipping it's build then

Screenshot 2024-06-27 at 9 11 55 PM
I lied big time, almost 19GB

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 13 above)
Comment (danielrhuynh) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:136]: Gotcha, added

File: modules/docker-compose.perception.yaml (hunk 6 above)
Comment (danielrhuynh) [reply modules/docker-compose.perception.yaml:100]: This is for depth estimation?

File: src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py (hunk 7 above)
Comment (danielrhuynh) [reply src/perception/lidar_object_detection/lidar_object_detection/lidar_object_detection/lidar_object_detection_node.py:15]: Hmm I remember this being required for the node to recognize openpcdet for some reason, any advice to get around it?

**Review by Edwardius — APPROVED** [review]
